src/Components/Blazor/Blazor/src/Hosting/WebAssemblyHost.cs

```diff
@@ -5,94 +5,166 @@
-                // as well as the Configure method.
-                var builder = new WebAssemblyBlazorApplicationBuilder(_scope.ServiceProvider);
-                startup.Configure(builder, _scope.ServiceProvider);
+            _disposed = true;
 
-                _renderer = await builder.CreateRendererAsync();
+            _renderer?.Dispose();
+
+            if (_scope is IAsyncDisposable asyncDisposableScope)
+            {
+                await asyncDisposableScope.DisposeAsync();
             }
-            catch
+            else
             {
-                _scope.Dispose();
-                _scope = null;
-
-                if (_renderer != null)
-                {
-                    _renderer.Dispose();
-                    _renderer = null;
-                }
+                _scope?.Dispose();
+            }
 
-                throw;
+            if (_services is IAsyncDisposable asyncDisposableServices)
+            {
+                await asyncDisposableServices.DisposeAsync();
+            }
+            else if (_services is IDisposable disposableServices)
+            {
+                disposableServices.Dispose();
             }
         }
 
-        public Task StopAsync(CancellationToken cancellationToken = default)
+        /// <summary>
+        /// Runs the application associated with this host. 
+        /// </summary>
+        /// <returns>A <see cref="Task"/> which represents exit of the application.</returns>
+        /// <remarks>
+        /// At this time, it's not possible to shut down a Blazor WebAssembly application using imperative code.
+        /// The application only stops when the hosting page is reloaded or navigated to another page. As a result
+        /// the task returned from this method does not complete. This method is not suitable for use in unit-testing.
+        /// </remarks>
+        public Task RunAsync()
         {
-            if (_scope != null)
+            // RunAsyncCore will await until the CancellationToken fires. However, we don't fire it
+            // currently, so the app will "run" forever.
+            return RunAsyncCore(CancellationToken.None);
+        }
+
+        // Internal for testing.
+        internal async Task RunAsyncCore(CancellationToken cancellationToken)
+        {
+            if (_started)
             {
-                _scope.Dispose();
-                _scope = null;
+                throw new InvalidOperationException("The host has already started.");
             }
 
-            if (_renderer != null)
+            _started = true;
+
+            var tcs = new TaskCompletionSource<object>();
+
+            using (cancellationToken.Register(() => { tcs.TrySetResult(null); }))
             {
-                _renderer.Dispose();
-                _renderer = null;
-            }
+                await using (this)
+                {
+                    var loggerFactory = (ILoggerFactory)Services.GetService(typeof(ILoggerFactory));
```

Comment:
```suggestion
                    var loggerFactory = Services.GetRequiredService<ILoggerFactory>();
```

Reply:
https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/RenderTree/Renderer.cs#L63-L66 will throw if the logger factory is null

Reply:
noice.